### PR TITLE
update addresses.json file using vm.writeJson

### PR DESCRIPTION
**Pull request by prat-gpt**
Issue: https://github.com/solidity-labs-io/forge-proposal-simulator/issues/78

### CODE REVIEW

File: addresses/Addresses.sol [hunk 1]
```diff
@@ -178,6 +185,16 @@ contract Addresses is IAddresses, Test {
...
+        for (uint256 i; i < savedAddresses.length; i++) {
+            if (
+                keccak256(abi.encode(savedAddresses[i].name)) ==
+                keccak256(abi.encode(name)) &&
+                savedAddresses[i].chainId == chainId
+            ) {
+                savedAddresses[i].addr = _addr;
+            }
+        }
+
```
Comment (ElliotFriedman) [inline addresses/Addresses.sol:197]: what is the purpose of adding this?

Comment (ElliotFriedman) [reply addresses/Addresses.sol:197]: ok, what you're saying is if you added this address, and then change it in the same proposal, this handles that case

File: test/Addresses.t.sol [hunk 2]
```diff
@@ -329,7 +329,7 @@ contract TestAddresses is Test {
             memory addressesPath = "./addresses/AddressesDuplicatedDifferentName.json";
 
         vm.expectRevert(
-            "Address: 0x9679e26bf0c470521de83ad77bb1bf1e7312f739 already set on chain: 31337"
+            "Address: 0x9679E26bf0C470521DE83Ad77BB1bf1e7312f739 already set on chain: 31337"
```
Comment (ElliotFriedman) [inline test/Addresses.t.sol:369]: would be great to see a unit test that adds an address, then writes it to JSON, then reads that JSON back in to see that the file was correctly written to

Comment (prat-gpt) [reply test/Addresses.t.sol:369]: added test here [35b8930](https://github.com/solidity-labs-io/forge-proposal-simulator/pull/84/commits/35b89303279cc21da8b394891e917a3ba93e07cb)

File: test/Addresses.t.sol [hunk 3]
```diff
@@ -275,6 +275,20 @@ contract TestAddresses is Test {
...
+
+        addresses.addAddress("TEST", test, 123, true);
+
+        addresses.updateJson();
+
+        string memory addressesPath = "./addresses/Addresses.json";
+        addresses = new Addresses(addressesPath);
+
+        // check Addresses.json is updated correctly and TEST address is set
+        addresses.isAddressSet("TEST");
```
Comment (ElliotFriedman) [inline test/Addresses.t.sol]: also add a getAddress to these tests

**Review by ElliotFriedman — CHANGES_REQUESTED** [review]
Would be great to see some docs here so that users can understand the new environment variables

File: docs/overview/architecture/proposal-functions.md [hunk 4]
```diff
@@ -287,7 +287,17 @@ FPS is flexible enough so that for any different governance model, governance pr
...
+
+    Flags used in `run()` function:
+
+    -   **DO_DEPLOY**: When set to true, triggers the deployment of contracts on-chain.
+    -   **DO_AFTER_DEPLOY_MOCK**: When set to true, initiates post-deployment mocking processes. Essential for testing or simulating scenarios after deployment.
+    -   **DO_BUILD**: When set to true, controls the build process and transforms plain solidity code into calldata encoded for the user's governance model.
+    -   **DO_SIMULATE**: When set to true, allows for the simulation of saved actions during the `build` step.
+    -   **DO_VALIDATE**: When set to true, validates the system state post-proposal simulation.
+    -   **DO_PRINT**: When set to true, prints proposal description, actions, and calldata.
+    -   **DO_UPDATE_ADDRESS_JSON**: When set to true, updates the `Addresses.json` file with the newly added and changed addresses.
```
Comment (ElliotFriedman) [inline docs/overview/architecture/proposal-functions.md]: Add default values to the docs

File: docs/overview/architecture/proposal-functions.md [hunk 5]
```diff
@@ -287,7 +287,17 @@ FPS is flexible enough so that for any different governance model, governance pr
...
+-   `run()`: This function serves as the entry point for proposal execution. It selects the `primaryForkId` which will be used to run the proposal simulation. It executes `deploy()`, `afterDeployMock()`, `build()`, `simulate()`, `validate()`, `print()` and `addresses.updateJson()` in that order if the flag for a function is set to true. `deploy()` is encapsulated in start and stop broadcast. This is done so that contracts can be deployed on-chain.
+
+    Flags used in `run()` function:
+
+    -   **DO_DEPLOY**: When set to true, triggers the deployment of contracts on-chain.
+    -   **DO_AFTER_DEPLOY_MOCK**: When set to true, initiates post-deployment mocking processes. Essential for testing or simulating scenarios after deployment.
```
Comment (ElliotFriedman) [inline docs/overview/architecture/proposal-functions.md]:
```suggestion
    -   **DO_AFTER_DEPLOY_MOCK**: When set to true, initiates post-deployment mocking processes. Used to simulate an action that has not happened yet for testing such as dealing tokens for testing or simulating scenarios after deployment.
```

File: src/proposals/Proposal.sol [hunk 6]
```diff
@@ -23,14 +23,21 @@ abstract contract Proposal is Test, Script, IProposal {
...
     bool internal DO_AFTER_DEPLOY_MOCK;
+    /// @notice flag to transform plain solidity code into calldata encoded for the user's governance model
     bool internal DO_BUILD;
+    /// @notice flag to simulate saved actions during the `build` step
     bool internal DO_SIMULATE;
+    /// @notice flag to validate the system state post-proposal simulation
     bool internal DO_VALIDATE;
+    /// @notice flag to print proposal description, actions, and calldata
     bool internal DO_PRINT;
+    /// @notice flag to update the `Addresses.json` file with the newly added and changed addresses
```
Comment (ElliotFriedman) [inline src/proposals/Proposal.sol]: add default values to natspec

File: test/Addresses.t.sol [hunk 7]
```diff
@@ -275,6 +275,20 @@ contract TestAddresses is Test {
         assertEq(addresses.isAddressContract("TEST"), true);
     }
 
+    function test_checkAddressFileUpdate() public {
+        address test = vm.addr(1);
+
+        addresses.addAddress("TEST", test, 123, true);
```
Comment (ElliotFriedman) [inline test/Addresses.t.sol]: let's do a more involved test than a single address being added

Mix it up a bit, here are some ideas:
- add two addresses
- add one address, then change where it points
- mix in writing to the file, then creating a new addresses object reading from that file, check isAddressSet and getAddress using both the chain id and without the chain id as parameters.

File: docs/overview/architecture/proposal-functions.md [hunk 8]
```diff
@@ -291,13 +291,13 @@ FPS is flexible enough so that for any different governance model, governance pr
...
     Flags used in `run()` function:
 
-    -   **DO_DEPLOY**: When set to true, triggers the deployment of contracts on-chain.
-    -   **DO_AFTER_DEPLOY_MOCK**: When set to true, initiates post-deployment mocking processes. Essential for testing or simulating scenarios after deployment.
-    -   **DO_BUILD**: When set to true, controls the build process and transforms plain solidity code into calldata encoded for the user's governance model.
-    -   **DO_SIMULATE**: When set to true, allows for the simulation of saved actions during the `build` step.
-    -   **DO_VALIDATE**: When set to true, validates the system state post-proposal simulation.
-    -   **DO_PRINT**: When set to true, prints proposal description, actions, and calldata.
-    -   **DO_UPDATE_ADDRESS_JSON**: When set to true, updates the `Addresses.json` file with the newly added and changed addresses.
+    -   **DO_DEPLOY**: When set to true, triggers the deployment of contracts on-chain. Default value is false.
```
Comment (ElliotFriedman) [inline docs/overview/architecture/proposal-functions.md]:
```suggestion
    -   **DO_DEPLOY**: When set to true, triggers the deployment of contracts on-chain. Default value is true.
```

**Review by ElliotFriedman — APPROVED** [review]
LGTM!

**Review by anajuliabit — COMMENTED** [review]
this is so cool, LGTM!

**Pull request merged**